### DISCUSSION TITLE
About headless mode...

### DISCUSSION THREAD

**Discussion by mgerhold** [mgerhold/silly-sprites#31] [General]
@kgnet88 provided some useful information on how it could be possible to run the engine in headless mode, but still provide OpenGL rendering (e.g. for unit tests). See their post [here](https://github.com/mgerhold/silly-sprites/issues/23#issuecomment-1652086463).

The main drawback I see here: Without providing (= paying for) a custom runner for our GitHub actions, we won't get any GPU support. Sure, there are projects that allow rendering OpenGL on a CPU, but I don't think this is worth the effort.

So I think: We should support headless mode (via a command line flag), but then the rendering system should be disabled alltogether. We can still support unit tests, but only for machines that have native GPU support and can open a window. I think that's okay.

**Comment by kgnet88** [comment]
If I understand you correctly, that means
- unit testing via hidden window (convenience) only on local machines
- headless server mode completely without graphics support
correct?

I'd love to disagree with you, but I think you're right and it's the best solution for what you discussed in the stream at the moment.

The only scenario for a headless server mode with graphics support that I can think of at the moment would be to outsource calculations to a network node in order to increase the simulation / render power of the system. This is extremely interesting for me in the real-time simulation area, but I can't think of a good scenario for a game engine right now (except maybe cloud gaming, like Nvidia Now does).

I will definitely implement a few examples with the EGL and test them via a server, if only to try out how easy it is to outsource a physical simulation and how much overhead the whole thing produces, but of course that is far outside of your goals...

**Reply by mgerhold** [reply]
I don't even know if we need support for the "hidden window" approach because the software creates a window anyway (regardless of whether it's hidden or not). So for unit testing on local machines, I think "normal rendering" would be feasable as well. The software would only need support for saving screenshots.

As you pointed out: The only real use cases for headless rendering are way beyond the scope of this project, so let's try to keep it simple(-ish) 😅
